### PR TITLE
[SPARK-35128][UI] Some columns in table Data Distribution of storage page shows incorrectly when sorted

### PR DESCRIPTION
**Pull request by echohlne**
### What changes were proposed in this pull request?
columns like 'On Heap Memory Useage', 'On Heap Memory Useage', 'Disk Usage' in table `Data Distribution` of storage page should be sorted as numerical-order instead of lexicographical-order.

The problem is very similar to #32190, but the cause of the problem is completely different, so a new pr was submitted.

### Why are the changes needed?
buf fix,the sorting style should be consistent between different columns.

When the storage page is initialized, it will infer the type of columns in the table to detect which sorting method(such as `sort_alpha`, `sort_numberic`, `sort_mmdd` etc.) to use. 

**Columns which contains a storage unit in the table will be inferred to be sorted use function `sort_alpha`, but the data (numerical value+unit)should be converted into the actual number of bytes and then sorted.**

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
Only JS was modified, and the manual test result works well.

before:
data distribution table in storage tab sort incorrect

after:
after modified

### CODE REVIEW

**Comment by kyoty** [comment]
cc @sarutak  @gengliangwang

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 1]
```diff
@@ -185,6 +185,10 @@ sorttable = {
         if (text.match(/^-?[£$¤]?[\d,.]+%?$/)) {
           return sorttable.sort_numeric;
         }
+        // match content like '287.1 MiB (17.2 GiB Remaining)', '0.0 B'
+        else if(text.match(/^(\d+(?:\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)[\s\S]*/i)) {
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:189]: `else` keyword should be immediately after `if` block with one whitespace inserted between `}` and `else` like:
```
if ( ... ) {
} else if ( ... ) {
  ...
}
```

Also, you should insert one whitespace between `if` and `(`.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 2]
```diff
@@ -278,6 +282,19 @@ sorttable = {
     if (isNaN(bb)) bb = 0;
     return aa-bb;
   },
+  sort_storage_size: function(src, dst) {
+    if(src[0].toLowerCase() === dst[0].toLowerCase()) {
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:286]: nit: Please insert one whitespace between `if` and `(`.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 3]
```diff
@@ -278,6 +282,19 @@ sorttable = {
...
+      return 0;
+    }
+    srcSize = storageSizeConverter(src[0])
+    dstSize = storageSizeConverter(dst[0])
+
+    if(srcSize === dstSize) {
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:292]: Please insert one whitespace.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 4]
```diff
@@ -278,6 +282,19 @@ sorttable = {
...
+      return 0;
+    }else {
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:294]: Please insert one whitespace.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 5]
```diff
@@ -505,3 +522,28 @@ var forEach = function(object, block, context) {
...
+    tib: 1099511627776,
+    pib: 1125899906842624,
+    eib: 1152921504606846976,
+    zib: 1180591620717411303424,
+    yib: 1208925819614629174706176,
+  }
+
+  var matches = data.match(/^(\d+(?:\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)[\s\S]*/i);
+
+  if(matches) {
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:542]: Please insert one whitespace.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 6]
```diff
@@ -505,3 +522,28 @@ var forEach = function(object, block, context) {
   }
 };
 
+// getStorageSize from str like '9.75 Mib'/'1290 B'
+function storageSizeConverter(data) {
+  var data =data.toString()
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:527]: Please insert one white space after `=`. and `;` after this statement.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 7]
```diff
@@ -278,6 +282,19 @@ sorttable = {
     if (isNaN(bb)) bb = 0;
     return aa-bb;
   },
+  sort_storage_size: function(src, dst) {
+    if(src[0].toLowerCase() === dst[0].toLowerCase()) {
+      return 0;
+    }
+    srcSize = storageSizeConverter(src[0])
+    dstSize = storageSizeConverter(dst[0])
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:290]: Please append `;` to the statements above.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 8]
```diff
@@ -185,6 +185,10 @@ sorttable = {
         if (text.match(/^-?[£$¤]?[\d,.]+%?$/)) {
           return sorttable.sort_numeric;
         }
+        // match content like '287.1 MiB (17.2 GiB Remaining)', '0.0 B'
+        else if(text.match(/^(\d+(?:\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)[\s\S]*/i)) {
+          return sorttable.sort_storage_size
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:190]: Please append `;` to this statement.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 9]
```diff
@@ -505,3 +522,28 @@ var forEach = function(object, block, context) {
...
+    b: 1,
+    kib: 1024,
+    mib: 1048576,
+    gib: 1073741824,
+    tib: 1099511627776,
+    pib: 1125899906842624,
+    eib: 1152921504606846976,
+    zib: 1180591620717411303424,
+    yib: 1208925819614629174706176,
+  }
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:538]: `;`

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 10]
```diff
@@ -505,3 +522,28 @@ var forEach = function(object, block, context) {
...
+
+  var matches = data.match(/^(\d+(?:\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)[\s\S]*/i);
+
+  if(matches) {
+    var unit = matches[2].toLowerCase()
```
Comment (sarutak) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:543]: `;`

**Comment by kyoty** [comment]
Thanks  for your review! @sarutak  I'll pay more attention to the code format.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 11]
```diff
@@ -184,6 +184,9 @@ sorttable = {
       if (text != '') {
         if (text.match(/^-?[£$¤]?[\d,.]+%?$/)) {
           return sorttable.sort_numeric;
+        } else if (text.match(/^(\d+(?:\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)[\s\S]*/i)) {
```
Comment (gengliangwang) [inline core/src/main/resources/org/apache/spark/ui/static/sorttable.js:187]: I think the following should simply work:
```
text.match(/^(\d+(\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)/i
```
The `?:` and `[\s\S]*` part is not necessary here.

Comment (kyoty) [reply core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: done this.

**Comment by kyoty** [comment]
@gengliangwang  @sarutak  Thanks for your review, the code has been updated, please have a look when your are free.

**Comment by kyoty** [comment]
@gengliangwang @sarutak Thanks for your review, the code has been updated, please have a look when your are free.

File: core/src/main/resources/org/apache/spark/ui/static/sorttable.js [hunk 12]
```diff
@@ -505,3 +521,27 @@ var forEach = function(object, block, context) {
...
+    pib: 1125899906842624,
+    eib: 1152921504606846976,
+    zib: 1180591620717411303424,
+    yib: 1208925819614629174706176,
+  };
+
+  var matches = data.match(/^(\d+(\.\d+)?)\s*(b|[k|m|g|t|p|e|z|y]ib)/i);
+
+  if (matches) {
+    var unit = matches[2].toLowerCase();
```
[the remark this answers is not in the document]
Comment (kyoty) [reply core/src/main/resources/org/apache/spark/ui/static/sorttable.js]: @gengliangwang  done.

**Comment by kyoty** [comment]
@sarutak  I have rebased to master and run dev/lint-js to keep style clean.

**Pull request closed without merging**